homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,378 @@
+
+    async def async_revert(self) -> None:
+        """Restore the entities touched by the schedule."""
+        if not self._affected_states:
+            return
+
+        entities = {}
+        for state in self._affected_states:
+            data = {**state.attributes}
+            data["state"] = state.state
+            entities[state.entity_id] = data
+
+        await self._hass.services.async_call(
+            "scene",
+            "apply",
+            service_data={"entities": entities},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler reverted scene: %s", self.entity_id)
+
+    async def async_trigger(self) -> None:
+        """Trigger the schedule's scene."""
+
+        @callback
+        def store_entity_if_in_context(
+            entity_id: str, old_state: State, new_state: State
+        ) -> None:
+            """Save prior states of an entity if it was triggered by this schedule."""
+            if new_state.context == self._context:
+                self._affected_states.append(old_state)
+
+        self._async_state_listener = async_track_state_change(
+            self._hass, MATCH_ALL, store_entity_if_in_context
+        )
+
+        await self._hass.services.async_call(
+            "scene",
+            "turn_on",
+            service_data={"entity_id": self.entity_id},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler triggered scene: %s", self.entity_id)
+
+
+class Schedule:
+    """A class to represent a schedule."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        end_datetime: Optional[datetime] = None,
+        recurrence: Optional[rrule] = None,
+    ):
+        """Initialize."""
+        self._async_revert_listener: Optional[Callable[..., Awaitable]] = None
+        self._async_trigger_listener: Optional[Callable[..., Awaitable]] = None
+        self._hass: HomeAssistant = hass
+        self._initial_instance_scheduled: bool = False
+        self._is_on: bool = True
+        self.active_instance: Optional[ScheduleInstance] = None
+        self.end_datetime: Optional[datetime] = end_datetime
+        self.entity_id: str = entity_id
+        self.recurrence: Optional[rrule] = recurrence
+        self.schedule_id: str = uuid4().hex
+        self.start_datetime: datetime = start_datetime
+
+        self.instance_duration: Optional[timedelta]
+        if self.end_datetime:
+            self.instance_duration = self.end_datetime - self.start_datetime
+        else:
+            self.instance_duration = None
+
+    def __str__(self) -> str:
+        """Define the string representation of this schedule."""
+        return (
+            f'<Schedule start="{self.start_datetime}" '
+            f'end="{self.end_datetime}" rrule="{self.recurrence}">'
+        )
+
+    @property
+    def active(self) -> bool:
```

Comment:
Shouldn't this method just be `return self.active_instance is not None`  ?

Reply:
Because right now you're checking if it could have an active instance, if it were to be turned on. What would be the use case for that ?

Reply:
I'm trying to think about what to do when HASS boots and we're in the middle of a schedule. I do agree that we could streamline it down to looking at `self.active_instance`.